File: app/src/main/java/com/esorokin/lantector/model/interactor/BaseInteractor.java
package com.esorokin.lantector.model.interactor;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class BaseInteractor {
	private final CompositeDisposable compositeDisposable;

	protected BaseInteractor() {
		compositeDisposable = new CompositeDisposable();
	}

	protected void autoDispose(Disposable disposable) {
		compositeDisposable.add(disposable);
	}

	public void onDestroy() {
		compositeDisposable.dispose();
	}
}

File: app/src/main/java/com/esorokin/lantector/model/interactor/BaseInteractor.kt
package com.esorokin.lantector.model.interactor

import com.esorokin.lantector.utils.ext.AutoDisposableContainer
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

open class BaseInteractor protected constructor() : Disposable, AutoDisposableContainer {
    private var disposed = false
    private val compositeDisposable: CompositeDisposable = CompositeDisposable()

    fun onDestroy() {
        //override
    }

    override fun autoDispose(disposable: Disposable) {
        compositeDisposable.add(disposable)
    }

    override fun isDisposed(): Boolean = disposed

    final override fun dispose() {
        disposed = true
        compositeDisposable.clear()
        onDestroy()
    }
}
